shapely/tests/test_set_operations.py

```diff
@@ -336,11 +342,34 @@ def test_coverage_union_overlapping_inputs():
     ],
 )
 def test_coverage_union_non_polygon_inputs(geom_1, geom_2):
-    # Non polygon geometries raise an error
-    with pytest.raises(
-        shapely.GEOSException, match="Unhandled geometry type in CoverageUnion."
-    ):
-        shapely.coverage_union(geom_1, geom_2)
+    if shapely.geos_version >= (3, 12, 0):
+
+        def effective_geom_types(geom):
+            if hasattr(geom, "geoms") and not geom.is_empty:
+                gts = set()
+                for geom in geom.geoms:
+                    gts |= effective_geom_types(geom)
+                return gts
+            return {geom.geom_type.lstrip("Multi").replace("LinearRing", "LineString")}
+
+        geom_types_1 = effective_geom_types(geom_1)
+        geom_types_2 = effective_geom_types(geom_2)
+        if len(geom_types_1) == 1 and geom_types_1 == geom_types_2:
+            with ignore_invalid():
+                # these show "invalid value encountered in coverage_union"
+                result = shapely.coverage_union(geom_1, geom_2)
```

Comment:
This is likely from GEOS, but not obvious. They can take a while to investigate, e.g. https://github.com/libgeos/geos/issues/515